src/main.cpp

```diff
@@ -3131,6 +3201,19 @@ bool ContextualCheckBlock(const CBlock& block, CValidationState& state, CBlockIn
     const int nHeight = pindexPrev == NULL ? 0 : pindexPrev->nHeight + 1;
     const Consensus::Params& consensusParams = Params().GetConsensus();
 
+    // Check that all transactions have the Overwinter flag set appropriately.
+    bool isOverwinterActive = NetworkUpgradeActive(nHeight, consensusParams, Consensus::UPGRADE_OVERWINTER);
+    BOOST_FOREACH(const CTransaction& tx, block.vtx) {
+        if (isOverwinterActive && !tx.fOverwintered) {
+            return state.DoS(100, error("%s: contains pre-overwinter tx", __func__),
+                            REJECT_INVALID, "cb-no-pre-overwinter-tx-when-active");
+        }
+        if (!isOverwinterActive && tx.fOverwintered) {
```

Comment:
Sanity-check: this is a new rejection rule that will be enabled on current mainnet (before Overwinter activation) when 1.1.0 is released, but it changes nothing because it is semantically equivalent to the `tx.nVersion < MIN_TX_VERSION` rejection rule.

Reply:
Right.. so I will remove this as it is redundant... and then gate the entire foreach loop so we are not introducing an Overwinter check (rejection rule) before Overwinter activates.  I'll also remove the test for `cb-no-pre-overwinter-tx-when-active`.

Reply:
Hmm, it occurred to me this morning that technically this isn't entirely redundant, because we've changed how we parse transactions, storing the sign bit in `tx.fOverwintered`. So the previous `tx.nVersion < MIN_TX_VERSION` rule is now semantically equivalent to `tx.fOverwintered || tx.nVersion < MIN_TX_VERSION`. Maybe we should change that rule instead?

Reply:
i.e. what we actually do is move the above rule over next to the existing rule and combine them.

Reply:
Or, take the `GetHeader()` function that I extracted in #2903 and have it return an `int32_t`, and then just use `tx.GetHeader() < MIN_TX_VERSION` here.

Reply:
Per my comment in https://github.com/zcash/zcash/pull/2925/files#r167231233 we cannot remove this rule, as it is not in fact redundant. We also need to re-check all consensus rules to ensure that the semantic change of `tx.nVersion` (to never have its sign bit set) doesn't alter any other consensus rules (I'm reasonably confident that `tx.nVersion < MIN_TX_VERSION` is the only one affected by this, but I want to be certain).

Reply:
@daira @str4d This is a good argument for having a control byte (with the msb set) so that the existing `nVersion` field does not change semantics.  Per comment here: https://github.com/zcash/zips/pull/133#issuecomment-361432773
We could still do this...